docs/src/references/agent/software-management.md

```diff
@@ -0,0 +1,332 @@
+---
+title: Software management
+tags: [Reference, Agent, Software Management]
+sidebar_position: 3
+---
+
+Thin-edge software management is implemented by the `software-update` operation,
+which gives the ability to:
+- manage software packages of different types on the same device 
+- fetch a pertinent subset of the software packages installed on a device
+- install, update and remove software packages on a device
+
+## MQTT API
+
+The `software-update` operation API follows the [generic thin-edge rules for operations](device-management-api.md):
+- The `te/<device-topic-id>/cmd/software-update` topic is used to publish the type of software packages
+  that can be managed on the device with the given topic identifier.
+- Each `te/<device-topic-id>/cmd/software-update/+` topic is dedicated to a software update command instance,
+  where all the subsequent states of this instance are published during its execution.
+- The workflow is [generic with `"init"`, `"executing"`, `"successful"` and `"failed"` statutes](references/agent/device-management-api.md/#operation-workflow).
+
+### Operation registration
+
+The registration message of the `software-update` operation on a device:
+- must provide a `supportedTypes` list of the types of software package that can be installed on this device (e.g. `["apt", "docker"]`)
+- can provide a description of the operation and of each supported package type.
+
+```sh te2mqtt formats="v1"
+tedge mqtt pub --retain 'te/device/child001///cmd/software-update' '{
+    "description": "Install, update and remove software packages",
+    "supportedTypes": [
+      "apt",
+      "docker",
+    ]
+}'
+```
+
+### `init` state
+
+A `software-update` command is defined by an `"updateList"` array giving the packages to install, update or remove.
+
+- If `"updateList"` field is not provided or empty, then
+  - the command must change nothing on the device
+  - and simply return the list of software currently installed.
```

Comment:
Currently, `tedge-agent` implements 2 operations: `software-list` and `software-update`.
Here, I propose to merge the two in a single  `software-update` used to query and update software.

Indeed, having two operations add some (little) complexities with no real benefits beyond mimicking c8y.
- If there are two operations, does it make sense for a device to support one but not the other?
- Have the list of supported package types to be same for the two operations?

Reply:
Indeed, a counter argument to my proposal. However, I void avoid to run a `sofware-list` concurrently with a `software-update` as the result would not be deterministic.

Reply:
I would also favour having independent `software-list` and `software-update` commands, as this would allow for a device to only support a software-list operation when user's want to manage their software via pure OS updates (for improved robustness, by having the rollback safety of A/B style updates).

Reply:
Okay. I will rewrite this proposal with two independent `software-list` and `software-update` commands.

Reply:
> Have the list of supported package types to be same for the two operations?

I'm of the opinion that the `device software list` and `updatable software list` would mostly likely be different for many customers. They may want to view all the installed packages from the cloud but limit the updatable list to a much smaller subset only for the ones that are safe to be updated remotely. Unfortunately C8y has them merged, for simplicity probably. I'd argue the same for configs as well. Not all readable configs are updatable.

Reply:
This has been addressed by https://github.com/thin-edge/thin-edge.io/pull/2235/commits/a48766243fe489a0bfdbe65b1022f145d007d1a2